usecases/scheduled_execution/batch_filtering.go

```diff
@@ -0,0 +1,185 @@
+	}
+
+	return models.Filter{
+		LeftSql:    leftVal.rawSql,
+		LeftValue:  leftVal.value,
+		Operator:   comparisonAstNodeToString(node),
+		RightSql:   rightVal.rawSql,
+		RightValue: rightVal.value,
+	}, true
+}
+
+func comparisonAstNodeToString(node ast.Node) string {
+	switch node.Function {
+	case ast.FUNC_GREATER:
+		return ">"
+	case ast.FUNC_GREATER_OR_EQUAL:
+		return ">="
+	case ast.FUNC_LESS:
+		return "<"
+	case ast.FUNC_LESS_OR_EQUAL:
+		return "<="
+	case ast.FUNC_EQUAL:
+		return "="
+	case ast.FUNC_NOT_EQUAL:
+		return "!="
+	default:
+		return ""
+	}
+}
+
+type parsedFilter struct {
+	rawSql          string
+	value           any
+	involvesPayload bool
+	valid           bool
+}
+
+func payloadFieldNodeToSql(node ast.Node, table models.TableIdentifier) (string, bool) {
+	field, ok := node.Children[0].Constant.(string)
+	return pgx.Identifier.Sanitize([]string{table.Schema, table.Table, field}), ok
+}
+
+func comparisonValueFromNode(node ast.Node, table models.TableIdentifier) parsedFilter {
+	// TODO: handle identifier sanitization against SQL injection
+
+	// case 1: payload field
+	if node.Function == ast.FUNC_PAYLOAD {
+		field, ok := payloadFieldNodeToSql(node, table)
+		return parsedFilter{rawSql: field, involvesPayload: true, valid: ok}
+	}
+
+	// case 2: date add operator
+	if node.Function == ast.FUNC_TIME_ADD {
+		var involvesPayload bool
+		timeNode := node.NamedChildren["timestampField"]
+		var time string
+		if timeNode.Function == ast.FUNC_TIME_NOW {
+			time = "now()"
+		} else if timeNode.Function == ast.FUNC_PAYLOAD {
+			field, ok := payloadFieldNodeToSql(timeNode, table)
+			if !ok {
+				return parsedFilter{}
+			}
+			involvesPayload = true
+			time = field
+		} else {
+			return parsedFilter{}
+		}
+
+		sign, ok := node.NamedChildren["sign"].Constant.(string)
+		if !ok || (sign != "+" && sign != "-") {
+			return parsedFilter{}
+		}
+		durationStr, ok := node.NamedChildren["duration"].Constant.(string)
+		if !ok {
+			return parsedFilter{}
+		}
+		if _, err := duration.Parse(durationStr); err != nil {
+			// check against possible SQL injection
+			return parsedFilter{}
+		}
+
+		return parsedFilter{
+			rawSql:          fmt.Sprintf("%s %s interval '%s'", time, sign, durationStr),
+			involvesPayload: involvesPayload,
+			valid:           true,
+		}
+	}
+
+	// case 3: now operator
+	if node.Function == ast.FUNC_TIME_NOW {
+		return parsedFilter{rawSql: "now()", valid: true}
+	}
+
+	// case 4: constant value
+	//   - string
+	//   - number
+	//   - boolean
+	//   - list of strings => should not happen here because it's a comparison operator, ignore it if it does anyway
+	if node.Constant != nil {
+		// switch on type
+		switch node.Constant.(type) {
+		case string, float64, bool:
+			return parsedFilter{value: node.Constant, valid: true}
+		default:
+			return parsedFilter{}
+		}
+	}
+
+	// return no filter in all other cases, in particular aggregates etc.
+	return parsedFilter{}
```

Comment:
And also, nesting !

Reply:
Daaamn, that's a good catch and one thing I forgot.
Ok, if I understand correctly what I should worry about is mostly the nested math operations: `+ - * /` (technically there are also all the comparison operators returning booleans, but perhaps I can ignore those for now since nested bool conditions seem unlikely).
Well, guess I'll have to handle those - protecting against any division by zero errors with NULLIF)

Reply:
Ok I have something that works with nested - cut at depth 3 max for safety